QPy Deserialization

amaloney opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This commit modifies the following modules.

- `gms.py`: The `MSGate` in this module states in its docstring that it has been deprecated in favor of the `GMS` class. The `GMS` docstring states that the parameter `theta` is a `num_qubits x num_qubits` matrix, and then has some logic to ensure the given theta is the correct shape. This logic has been refactored to fix the failing example given in Issue #11378.
- `circuits.py`: The example in Issue #11378 needed extra logic to handle an `MS` gate during [de]serialization. The refactor adds this logic, and switches the deprecated `MSGate` for the `GSM` one.

### Details and comments

The MRE for the error is in the issue, and replicated below.

```python
import io
from qiskit import qiskit, qpy

circuit = qiskit.QuantumCircuit(2)
circuit.ms(1.23, [0, 1])
buf = io.BytesIO()
qpy.dump(circuit, buf)
_ = qpy.load(io.BytesIO(buf.getvalue()))
```

The changes fixes the following error

```python
TypeError: MSGate.__init__() missing 1 required positional argument: 'theta'
```

Resolves #11378

ndmlny-rs committed:
QPy Deserialization

This commit modifies the following modules.

- `gms.py`: The `MSGate` in this module states in its docstring that it
  has been deprecated in favor of the `GMS` class. The `GMS`
  docstring states that the parameter `theta` is a `num_qubits x
  num_qubits` matrix, and then has some logic to ensure the given theta
  is the correct shape. This logic has been refactored to fix the failing
  example given in Issue #11378.
- `circuits.py`: The example in Issue #11378 needed extra logic to
  handle an `MS` gate during [de]serialization. The refactor adds this
  logic, and switches the deprecated `MSGate` for the `GSM` one.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=11506) <br/>All committers have signed the CLA.

amaloney commented:
Converting to `WIP` status, as I need to incorporate what @jakelishman explained in issues #11504 and #11507 for the local dev environment, and running tests locally.

kataro92 commented on qiskit/qpy/binary_io/circuits.py:
```diff
@@ -347,6 +347,11 @@ def _read_instruction(
                     gate.label = label
+            # NOTE: MSGate is deprecated, so we will switch it to the GMS gate. See
+            #       https://github.com/Qiskit/qiskit/issues/11378.
+            if gate_name == "MSGate":
```
using `if` in here change the logic of this condition.
The else in Line355 now will be same as `if label is not None and gate_name != "MSGate"`
In my opinion, you should use `elif`
My proposal code for Line352: `elif gate_name == "MSGate":`

amaloney committed:
Rebase & comment resolution

Rebases against main, and resolves the review comment.

amaloney commented:
@kataro92 my appologies for having 2 different users that are mine on this PR. I made sure to sign the CLA for both users, and will not make that mistake again.

1ucian0 reviewed (commented):
Dont forget to add a test and a release note!

amaloney commented:
thx @1ucian0 I think I got the pattern for PRs down a bit better with the #12009 PR and @jakelishman helping, and I'll update this one with that pattern.

Thanks again for helping as I get better with the codebase.

amaloney committed:
Add test and release note

- This commit adds a release note and a test.
- Tests were done with qiskit version 1.1

amaloney committed:
fix lint issue

1ucian0 commented on test/python/qpy/test_io_from_qpy.py:
you can add the test to test/python/qpy/test_circuit_load_from_qpy.py instead of creating a new file. I think you could reuse `assert_roundtrip_equal` there.

amaloney replied:
Running the tests using the `assert_roundrtip_equal` fails during the check. The reason for this is that the circuits are being modified from a `MSGate` to a `GSM` one since the `MSGate` has been deprecated. This modification was previously not communicated to the user, so I have created a `FutureWarning` when they serializes a circuit using QPY that uses a `MSGate`. 

The test needs to be updated to trigger the `FutureWarning` and ensure the circuits are equal, but I've left a comment to discuss design around the `MSGate` and `GSM` as the simple fix may cause unintended user confusion.

1ucian0 commented on releasenotes/notes/qpy-serialization-f08c241209706da2.yaml:
```diff
@@ -0,0 +1,10 @@
+    `qiskit.qpy.dump` and `qiskit.qpy.load`. Previously, an
+    error was thrown requiring a positional argument `theta` that is
+    now computed generated in the gate method.
+upgrade_qpy:
+  - |
+    Works in 1.x and may need a backport for 0.4x.
```
```suggestion
```

I will tag the PR as a candidate for 0.46 patch. This release note does not go here.

amaloney replied:
Understood, thanks for the clarification

1ucian0 commented on releasenotes/notes/qpy-serialization-f08c241209706da2.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+features_qpy:
+  - |
+    qiskit.qpy now saves and loads circuits reliably using
+    `qiskit.qpy.dump` and `qiskit.qpy.load`. Previously, an
```
```suggestion
    The QPY module (:mod:`qiskit.qpy`) now saves and loads circuits with reliably :class:`qiskit.circuit.library.MSGate` gates
    :func:`~qiskit.qpy.dump` and :func:`~qiskit.qpy.dump`. Previously, an
```

amaloney replied:
excellent suggestion, updated to the better language and links

1ucian0 commented on test/python/qpy/test_io_from_qpy.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def test_io():
+    """Test QPy serialization and de-serialization."""
```
It's not uncommon to refer to the issue in a regression test.
```suggestion
    """Test QPY serialization and de-serialization.
    See: https://github.com/Qiskit/qiskit/issues/11378"""
```

amaloney replied:
Updated the docstring for the test, that now lives in `test/python/qpy/test_circuit_load_from_qpy.py`

amaloney commented:
I'm not sure what the best course of action for serialization/deserialization using QPY is, since the `MSGate` has been deprecated in favor for the `GSM` gate, and the fix changes any MSGate to a GSM one. I think a simple thing to do is to put the theta checks done in the `GSM` method in the `MSGate` one as well. That way we can keep the `MSGate` in the ser-de of the circuit, but still issue a future warning that the gate has been deprecated and that a user should use the GSM one in the future.

@1ucian0 I'm open to design decisions that I can implement to make this easier for the user, and any comment you may have about the above suggestion.

1ucian0 commented:
@mtreinish any advice here?

KevinT3Hu commented:
What is the status on this, I encounter this error on qiskit 1.4.1 now.

gadial commented:
Thank you for this PR. In Qiskit 2.4 we moved the QPY framework to Rust, which means this PR is no longer relevant (however, the error persists; we should fix it in the new module). As a result, I'm closing this PR.
